airbyte-integrations/connectors/destination-iceberg-v2/src/main/kotlin/io/airbyte/integrations/destination/iceberg/v2/IcebergSuperTypeFinder.kt

```diff
@@ -0,0 +1,137 @@
+        // If the two types are already deeply equal, return one of them (arbitrary).
+        if (icebergTypesComparator.typesAreEqual(incomingType, existingType)) {
+            return existingType
+        }
+        // Otherwise, attempt to combine them into a valid supertype.
+        return combineTypes(existingType, incomingType, columnName)
+    }
+
+    /**
+     * Combines two top-level [Type]s. If exactly one is primitive and the other is non-primitive,
+     * no supertype is possible => throws [IllegalArgumentException].
+     */
+    private fun combineTypes(existingType: Type, incomingType: Type, columnName: String): Type {
+        if (existingType.isPrimitiveType != incomingType.isPrimitiveType) {
+            throwIllegalTypeCombination(existingType, incomingType, columnName)
+        }
+
+        // Both are primitive
+        if (existingType.isPrimitiveType && incomingType.isPrimitiveType) {
+            return combinePrimitives(
+                existingType.asPrimitiveType(),
+                incomingType.asPrimitiveType(),
+                columnName
+            )
+        }
+
+        // Both are non-primitive => not currently supported
+        throwIllegalTypeCombination(existingType, incomingType, columnName)
+    }
+
+    /**
+     * Checks whether either type is unsupported or unmapped (e.g. BINARY, DECIMAL, FIXED, etc.).
+     *
+     * @throws IllegalArgumentException if either type is unsupported.
+     */
+    private fun validateTypeIds(typeId1: TypeID, typeId2: TypeID) {
+        val unsupportedTypeIds = setOf(BINARY, DECIMAL, FIXED, UUID, MAP, TIMESTAMP_NANO)
+        if (typeId1 in unsupportedTypeIds || typeId2 in unsupportedTypeIds) {
+            val badTypeId = if (typeId1 in unsupportedTypeIds) typeId1 else typeId2
+            throw IllegalArgumentException(
+                "Unsupported or unmapped Iceberg type: $badTypeId. Please implement handling if needed."
+            )
+        }
+    }
+
+    /**
+     * Attempts to combine two [PrimitiveType]s into a valid supertype by using
+     * [TypeUtil.isPromotionAllowed].
+     *
+     * - If they have the same [TypeID], just returns the existing type (since they’re not deeply
+     * equal, but the top-level ID is the same. You may want to consider e.g. TIMESTAMP with/without
+     * UTC).
+     * - If they have different IDs, tries known promotions (INT->LONG, FLOAT->DOUBLE).
+     * - If promotion is not allowed, throws [IllegalArgumentException].
+     */
+    private fun combinePrimitives(
+        existingType: PrimitiveType,
+        incomingType: PrimitiveType,
+        columnName: String
+    ): Type {
+        val existingTypeId = existingType.typeId()
+        val incomingTypeId = incomingType.typeId()
+
+        validateTypeIds(existingTypeId, incomingTypeId)
+
+        // If promotion is not allowed by Iceberg, fail fast.
+        if (!TypeUtil.isPromotionAllowed(existingType, incomingType)) {
```

Comment:
Should this be happening even before `validateTypeIds`?

Reply:
done